SAK-49287 Notifications. Improved push permissions workflow

adrianfish opened the pull request:
https://sakaiproject.atlassian.net/browse/SAK-49287

adrianfish commented:
<img width="510" alt="Screenshot 2023-09-27 at 12 39 02" src="https://github.com/sakaiproject/sakai/assets/134546/764f14a8-3218-40ce-a8fa-4509e7bd4574">
<img width="856" alt="Screenshot 2023-09-27 at 12 36 49" src="https://github.com/sakaiproject/sakai/assets/134546/8b660447-eb10-47fb-8f2b-9ceb587254ad">
<img width="329" alt="Screenshot 2023-09-27 at 12 36 17" src="https://github.com/sakaiproject/sakai/assets/134546/96d0c845-032f-4a9b-bce3-1206f31369fd">

csev commented:
Tasty - I was wondering how we were going to handle this.   Folks say 'no' to the "accept notifications" for all sites - so for an LMS they need a second chance.  We will need to see if browsers kind of beat us up - they probably have some trick to keep sites from asking over and over.

ottenhoff commented on webcomponents/bundle/src/main/bundle/sui-notifications.properties:
```diff
@@ -13,6 +13,11 @@ mark_all_viewed=Mark all as viewed
 motd=Message of the Day
 no_notifications=No notifications
+notifications_not_allowed=You need to accept notifications from your Sakai server.
+notifications_not_allowed2=If you don't, you will not get any notifications regarding \
```
```suggestion
notifications_not_allowed2=If you do not accept, you will not receive notifications regarding \
```

ottenhoff commented on webcomponents/bundle/src/main/bundle/sui-notifications.properties:
```diff
@@ -13,6 +13,11 @@ mark_all_viewed=Mark all as viewed
+notifications_not_allowed=You need to accept notifications from your Sakai server.
+notifications_not_allowed2=If you don't, you will not get any notifications regarding \
+announcements, assignments, etc.
+notifications_denied=Notifications have been denied at some point and you will need to use your \
```
```suggestion
notifications_denied=Notifications were previously denied, and you will need to use your \
```

ern commented on sakai-root/pom.xml:
```diff
@@ -0,0 +1,38 @@
+
+  <name>Sakai Root</name>
+  <groupId>org.sakaiproject</groupId>
+  <artifactId>ROOT</artifactId>
```
Adrian can you explain this artifact ROOT? do we have a better name?

adrianfish replied:
So the war file is called ROOT.war and is deployed to the tomcat root, that's all.

adrianfish replied:
Do you think I can change the artifact id while still creating ROOT.war to deploy to tomcat?

adrianfish replied:
the reason we have this sakai-root thing is that we need to deploy a single service worker to the top level of our hierarchy so it can control all of Sakai's paths. It's crucial.

adrianfish replied:
the artifact is now "serviceworker" and I've used deployId to deploy to tomcat as the ROOT war.

ern reviewed (changes requested):
Not particularly a fan of the path names maybe we can discuss that?

adrianfish commented:
which path names in particular?

adrianfish commented:
<img width="483" alt="Screenshot 2024-01-31 at 18 14 56" src="https://github.com/sakaiproject/sakai/assets/134546/750cd6c2-b406-48c2-854c-735ed7af90a3">

adrianfish commented:
I've updated this to change the ROOT sakai maven module to just be called "root" and not "sakai-root". Also, I've added a button in the notifications window to make it explicit that the user is opting into Sakai notifications.

adrianfish commented:
I've now updated this maven module to be called "serviceworker". No more root.

stetsche commented:
@adrianfish I think I can take a look at the PR sometime next week, if it's still needed. But just to get some context, when the user does not accept the push notifications, there will be no notifications within the portal neither?

adrianfish commented:
@stetsche Exactly that. Our challenge is inform the user that Sakai notifications are good. Push is so powerful and it is a shame that companies have abused it so heavily.

stetsche commented:
@adrianfish, okay I understand, what the PR is for. But when I enable push notifications, will notifications display only in the portal, or like other push notifications as OS notifications (What a lot of websites are abusing)?

adrianfish commented:
Depending on the browser, both. Chrome and firefox will popup a OS notification as well as adding a notification to the portal area.

stetsche commented:
Okay, and if you want to have only portal notifications you need to enable them in the browser and block browser notifications on the OS level?

adrianfish commented:
I suppose so. Or we could just disable them in the js. They are explicitly triggered, the os notifications.

jesusmmp commented on webcomponents/tool/src/main/frontend/packages/sakai-push-utils/test/sakai-push-utils.test.js:
```diff
@@ -0,0 +1,18 @@
+
+  window.top.portal = { locale: 'en_GB' };
+
+  /*
```
Remove comment

ottenhoff commented on serviceworker/pom.xml:
```diff
@@ -0,0 +1,45 @@
+  <parent>
+    <groupId>org.sakaiproject</groupId>
+    <artifactId>master</artifactId>
+    <version>24-SNAPSHOT</version>
```
```suggestion
    <version>25-SNAPSHOT</version>
```

adrianfish commented:
Done with this now. I've removed sakai-message-broker.js. Most of the logic moved nicely into sakai-push.utils.js and the remainder I just tacked onto portal.init.js.

ottenhoff commented:
This seems to consistently break Cypress tests. I saw lots of errors in my console around theme switching. Maybe theme switcher was depending on something that got moved?

stetsche commented on webcomponents/tool/src/main/frontend/packages/sakai-push-utils/package.json:
```diff
@@ -0,0 +1,42 @@
+  "dependencies": {
+    "@sakai-ui/sakai-portal-utils": "file:../sakai-portal-utils",
+    "get-browser-fingerprint": "^3.2.0",
+    "lit": "^3.1.0"
```
lit? Could you review the dependencies?

stetsche commented on webcomponents/tool/src/main/frontend/packages/sakai-push-utils/src/sakai-push-utils.js:
```diff
@@ -0,0 +1,209 @@
+import getBrowserFingerprint from "get-browser-fingerprint";
```
where is this dependency defined?

stetsche commented on webcomponents/tool/src/main/frontend/packages/sakai-push-utils/README.md:
````diff
@@ -0,0 +1,47 @@
+To scan the project for linting and formatting errors, run
+
+```bash
+npm run lint
````
A lot of this scripts do not exits. Could you clean this up?

stetsche commented on webcomponents/tool/src/main/frontend/packages/sakai-push-utils/test/sakai-push-utils.test.js:
```diff
@@ -0,0 +1,8 @@
+import { pushSetupComplete } from '../src/sakai-push-utils.js';
+
+describe("sakai-push-utils tests", () => {
+
+  it ("sets up uccessfully", async () => {
+    pushSetupComplete.then(() => console.log("push setup complete"));
```
```suggestion
describe("sakai-push-utils", () => {

  it ("sets up successfully", async () => {
    pushSetupComplete.then(() => console.log("push setup complete"));
```

stetsche replied:
Though I am not sure if this test is asserting anything?

adrianfish replied:
No, it's not. It's just a sanity check at the minute. The test will fail if the imports fail etc. I need to add some more meaningful testing although it will have to involve service worker installation etc. So ... complex. I thought it was better to at least leave the shell of the test there.

stetsche commented on webcomponents/tool/src/main/frontend/packages/sakai-push-utils/web-test-runner.config.mjs:
```diff
@@ -0,0 +1,24 @@
+const filteredLogs = ['Running in dev mode', 'lit-html is in dev mode'];
```
Could this be trimmed as well?

adrianfish replied:
I think we need that. Or, do you mean it just needs updating to catch "lit is in dev mode" or something?

stetsche replied:
I said it not being sure what is needed and what not, I would be in favor of reducing the boiler plate code to a minimum. But you know better what is needed.

adrianfish replied:
I'm not sure if I know better or not ...

I know that I used to see these warnings about lit running in dev mode, and now I don't. I assume it's that filterLogs statement that has removed that warning when running tests.

stetsche reviewed (changes requested):
@adrianfish I can take a closer look at next week

stetsche commented on kernel/api/src/main/java/org/sakaiproject/messaging/api/model/UserNotification.java:
```diff
@@ -92,4 +92,7 @@ public class UserNotification implements PersistableEntity<Long> {
     private String tool;
+
+    @Transient
+    private Boolean isNotification = Boolean.TRUE;
```
Can UserNotification be anything else than an notification? This seems quite confusing to me. Maybe the variable name could be different or a comment added?

adrianfish replied:
Yes, it can in fact. A UserNotification can actually just be a message directly to a tool's js. So, not intended to be a popup type thing or even go into the Notifications area. It now seems a bad name, I agree.

stetsche replied:
Could you expand a bit on that use-case? It seems in this PR you don't make use of this feature. To me it's a bit odd, that we would update tool js when the user might not even have Sakai open.

adrianfish replied:
I've removed it. I added it to just make it clear in future that push is not just about visible notifications in the notifications area. The use case is this: somebody creates a new topic in conversations, the conversations service calls userMessagingService.message with a push message containing the new topic, or a reference to it, and that message is broadcast to all the relevant push subscriptions. The conversations js registers a push callback for those conversations messages and updates the client side ui in real time.

adrianfish replied:
We'd probably add a presence check for the tool messages. Think of tool messages as strictly online messages. Or, they could even be sent at all times and the service worker caches them. Anyway, some more thought is needed on that front, but live updating of tools is desirable.

adrianfish replied:
I'm working on another branch that starts the process of add progressive web app capabilities, based on this work.

stetsche replied:
I'm a bit skeptical regarding this approach, but it might be that I just don't know enough about this the push api. If you have any resources regarding push messages that are not notifications and best practices around it, it would be great if you could share them with me, I didn't find much about it. What you describe sounds like a good use-case for websockets to mee, but they might be outdated?
You can resolve this comment though.

adrianfish replied:
websockets are not suitable for our use. They keep a connection open so are more resource intensive on our servers.

adrianfish replied:
https://developer.mozilla.org/en-US/docs/Web/API/Push_API

The beauty of push is that we don't have to worry about how the message is delivered to the subscribed device. The vendor's push service deals with that - all we do is send the encrypted message to the push service and forget about it. Very scalable from our point of view.

stetsche replied:
Thanks!

adrianfish commented on webcomponents/bundle/src/main/bundle/sakai-notifications.properties:
```diff
@@ -15,10 +16,12 @@ message_read=Has read the message "{0}" from the site "{1}"
 motd=Message of the Day
 no_notifications=No notifications
+notifications_not_allowed=You need to accept notifications from your Sakai server.
+notifications_not_allowed2=If you don't, you will not get any notifications regarding \
```
I'm not sure about the regarding bit. That list is going to change, probably quite rapidly. In fact, we'll start using push to dynamically update tool markup, without a notification.

stetsche replied:
I would just suggest to replace the part "If you don't, you will not get any notifications regarding". To me the message sounds a bit like you're blackmailing the user to accept the notifications. Just as a minor suggestion though.

adrianfish replied:
I am blackmailing them :)
We really need them to accept this, or they will not get notifications. I want to make it really clear what the consequences are if they don't give permission.

stetsche replied:
If you insist :)

adrianfish commented on webcomponents/tool/src/main/frontend/packages/sakai-push-utils/src/sakai-push-utils.js:
```diff
@@ -0,0 +1,209 @@
+// the service worker first. We capture the registration in the closure.
+export const callSubscribeIfPermitted = async () => {
+
+  const reg = await navigator.serviceWorker.register("/sakai-service-worker.js");
```
sakai-service-worker is just the js file in the root, containing the service worker code.

adrianfish commented on portal/portal-render-engine-impl/impl/src/webapp/vm/morpheus/includeAccountNav.vm:
```diff
@@ -85,8 +85,12 @@ For opening help sidebar
-        <span class="portal-notifications-indicator p-1 bg-danger rounded-circle" style="display: none">
+        <i class="portal-notifications-icon bi-bell"></i>
+        <span class="portal-notifications-no-permissions-indicator p-1 bg-danger rounded-circle d-none">
+          X
```
I tried and it was a pain in the culo.

stetsche replied:
I hope the pain went away and we are left with a nice indicator : ) Code looks good

adrianfish replied:
Yes, the pain has gone.

stetsche commented on webcomponents/bundle/src/main/bundle/sakai-notifications.properties:
```diff
@@ -15,10 +16,12 @@ message_read=Has read the message "{0}" from the site "{1}"
 message_received={0} sent you a message
 motd=Message of the Day
 no_notifications=No notifications
+notifications_not_allowed=You need to accept notifications from your Sakai server.
```
I was thinking of the `ui.service` property. Could we use it?

ottenhoff commented on portal/portal-render-engine-impl/impl/src/webapp/vm/morpheus/includeAccountNav.vm:
```diff
@@ -85,9 +85,12 @@ For opening help sidebar
-          <span class="visually-hidden">There are new notifications available</span>
+        <i class="portal-notifications-icon bi-bell"></i>
+        <i class="portal-notifications-no-permissions-indicator si si-warning d-none"
+           aria-label="$rloader.sit_notifications_not_permitted_label">
```
aria-hidden?

adrianfish replied:
don't we want that aria-label read out by screen readers?

adrianfish replied:
Or does it not work like that? Will the screen reader read the tag and the label out? Unless we use aria-hidden as well?

ottenhoff replied:
you're exactly correct. sorry

stetsche commented on webcomponents/tool/src/main/frontend/packages/sakai-push-utils/src/sakai-push-utils.js:
```diff
@@ -0,0 +1,211 @@
+
+      try {
+        Notification.requestPermission().then(permission => {
+
+          console.log(permission);
```
```suggestion
```

stetsche reviewed (approved):
The PR looks good to me now. Well done @adrianfish! I only found a `console.log` to remove, but I'll mark this as approved already, so I don't have to come back to the PR

adrianfish commented:
Thanks a ton @stetsche, for the review. And anybody else who reviewed this PR. Much appreciated :)

bgarciaentornos commented:
I think the husky command is failing for windows
